### PR TITLE
Improve I/O BBEs to use new high-level APIs

### PR DESCRIPTION
**Pull request by BuddhiWathsala**
## Purpose
Fixes https://github.com/ballerina-platform/ballerina-standard-library/issues/573

## Approach
Use the new high-level IO APIs.

## Related PRs
https://github.com/ballerina-platform/module-ballerina-io/pull/31

## Test environment
- macOS
- Java11
-SLP6

### CODE REVIEW

**Review by lafernando — CHANGES_REQUESTED** [review]
Let's hold this a bit, until we finalize the new BBE structure.

**Comment by BuddhiWathsala** [comment]
Here only added the File I/O APIs since socket module implementation is not going to use I/O stdlib channel APIs.

**Review by lafernando — APPROVED** [review]

File: examples/file-io/file_io.bal [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+public function main() returns @tainted error? {
+    // Initialize the image paths.
+    string imagePath = "./files/ballerina.jpg";
+    string imageCopyPath1 = "./files/ballerinaCopy1.jpg";
+    string imageCopyPath2 = "./files/ballerinaCopy1.jpg";
+
+    // Reads the file content as a byte array using the given file path.
+    byte[] bytes = check io:fileReadBytes(imagePath);
+    // The content that already read, will write to the given destination file.
```
Comment (praneesha) [inline examples/file-io/file_io.bal:11]:
```suggestion
    // Write the content is that already read to the given destination file.
```

File: examples/file-io/file_io.bal [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    check io:fileWriteBytes(imageCopyPath1, bytes);
+    io:println("Successfully copied the image as a byte array.");
+
+    // Reads the file as a stream of blocks. The default block size is 4KB. Here, the default size is overridden with 2KB.
```
Comment (praneesha) [inline examples/file-io/file_io.bal:15]:
```suggestion
    // Read the file as a stream of blocks. The default block size is 4KB. Here, the default size is overridden by the value 2KB.
```

File: examples/file-io/file_io.bal [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    stream<io:Block> blockStream = check io:fileReadBlocksAsStream(imagePath, 2048);
+    // If the file reading was successful, then the content will write to the given destination file using the given stream.
```
Comment (praneesha) [inline examples/file-io/file_io.bal:17]:
```suggestion
    // If the file reading was successful, then, the content will be written to the given destination file using the given stream.
```

File: examples/file-io/file_io.bal [hunk 4]
```diff
@@ -0,0 +1,97 @@
+import ballerina/io;
+
+public function main() returns @tainted error? {
+    // Initialize the image paths.
+    string imagePath = "./files/ballerina.jpg";
+    string imageCopyPath1 = "./files/ballerinaCopy1.jpg";
+    string imageCopyPath2 = "./files/ballerinaCopy1.jpg";
+
+    // Reads the file content as a byte array using the given file path.
```
Comment (praneesha) [inline examples/file-io/file_io.bal:9]:
```suggestion
    // Read the file content as a byte array using the given file path.
```

File: examples/file-io/file_io.bal [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    io:println("Successfully copied the image as a stream.");
+
+    // Initialize the text path and the content.
+    string textFilePath1 = "./files/textfile1.txt";
+    string textFilePath2 = "./files/textfile2.txt";
+    string textFilePath3 = "./files/textfile3.txt";
+    string textContent = "Ballerina is an open source programming language.";
+    string[] lines = ["The Big Bang Theory", "F.R.I.E.N.D.S", "Game of Thrones", "LOST"];
+
+    // Writes the given string to a file.
```
Comment (praneesha) [inline examples/file-io/file_io.bal:28]:
```suggestion
    // Write the given string to a file.
```

File: examples/file-io/file_io.bal [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    check io:fileWriteString(textFilePath1, textContent);
+    // If the write operation was successful, then read the content as a string.
+    string readContent = check io:fileReadString(textFilePath1);
+    io:println(readContent);
+
+    // Writes the given array of lines to a file.
```
Comment (praneesha) [inline examples/file-io/file_io.bal:34]:
```suggestion
    // Write the given array of lines to a file.
```

File: examples/file-io/file_io.bal [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+    check io:fileWriteLines(textFilePath2, lines);
+    // If the write operation was successful, then perform a read operation to read the lines as an array.
+    string[] readLines = check io:fileReadLines(textFilePath2);
+    io:println(readLines);
+
+    // Writes the given stream of lines to a file.
```
Comment (praneesha) [inline examples/file-io/file_io.bal:40]:
```suggestion
    // Write the given stream of lines to a file.
```

File: examples/file-io/file_io.bal [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    check io:fileWriteLinesFromStream(textFilePath3, lines.toStream());
+    // If the write operation was successful, then perform a read operation to read the lines as a stream.
+    stream<string> lineStream = check io:fileReadLinesAsStream(textFilePath3);
+    // Loop through the stream and print the content.
+    _ = lineStream.forEach(function(string val) {
+                               io:println(val);
+                           });
+
+    // Initialize JSON file path and content.
```
Comment (praneesha) [inline examples/file-io/file_io.bal:49]:
```suggestion
    // Initialize the JSON file path and content.
```

File: examples/file-io/file_io.bal [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+            "@id": "AST",
+            "name": "Anne",
+            "address": {
+                "street": "Main",
+                "city": "94"
+            },
+            "codes": ["4", "8"]
+        }};
+
+    // Writes the given JSON to a file.
```
Comment (praneesha) [inline examples/file-io/file_io.bal:61]:
```suggestion
    // Write the given JSON to a file.
```

File: examples/file-io/file_io.bal [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+    check io:fileWriteJson(jsonFilePath, jsonContent);
+    // If the write operation was successful, then perform a read operation to read the JSON content.
+    json readJson = check io:fileReadJson(jsonFilePath);
+    io:println(readJson);
+
+    // Initialize XML file path and content.
```
Comment (praneesha) [inline examples/file-io/file_io.bal:67]:
```suggestion
    // Initialize the XML file path and content.
```

File: examples/file-io/file_io.bal [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+    string xmlFilePath = "./files/xmlFile.xml";
+    xml xmlContent = xml `<book>The Lost World</book>`;
+
+    // Writes the given XML to a file.
```
Comment (praneesha) [inline examples/file-io/file_io.bal:71]:
```suggestion
    // Write the given XML to a file.
```

File: examples/file-io/file_io.bal [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+    check io:fileWriteXml(xmlFilePath, xmlContent);
+    // If the write operation was successful, then perform a read operation to read the XML content.
+    xml readXml = check io:fileReadXml(xmlFilePath);
+    io:println(readXml);
+
+    // Initialize CSV file path and content.
```
Comment (praneesha) [inline examples/file-io/file_io.bal:77]:
```suggestion
    // Initialize the CSV file path and content.
```

File: examples/file-io/file_io.bal [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+    string csvFilePath1 = "./files/csvFile1.csv";
+    string csvFilePath2 = "./files/csvFile2.csv";
+    string[][] csvContent = [["1", "James", "10000"], ["2", "Nathan", "150000"], ["3", "Ronald", "120000"], ["4", "Roy",
+    "6000"], ["5", "Oliver", "1100000"]];
+
+    // Writes the given content string[][] to a CSV file.
```
Comment (praneesha) [inline examples/file-io/file_io.bal:83]:
```suggestion
    // Write the given content string[][] to a CSV file.
```

File: examples/file-io/file_io.bal [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+    check io:fileWriteCsv(csvFilePath1, csvContent);
+    // If the write operation was successful, then perform a read operation to read the CSV content as a string array of arrays.
+    string[][] readCsv = check io:fileReadCsv(csvFilePath1);
+    io:println(readCsv);
+
+    // Writes the given content stream to a CSV file.
```
Comment (praneesha) [inline examples/file-io/file_io.bal:89]:
```suggestion
    // Write the given content stream to a CSV file.
```

File: examples/file-io/file_io.bal [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+    check io:fileWriteCsvFromStream(csvFilePath2, csvContent.toStream());
+    // If the write operation was successful, then perform a read operation to read the CSV content as a stream.
```
Comment (praneesha) [inline examples/file-io/file_io.bal:91]:
```suggestion
    // If the write operation was successful, then, perform a read operation to read the CSV content as a stream.
```

File: examples/file-io/file_io.bal [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+
+    // Initialize CSV file path and content.
+    string csvFilePath1 = "./files/csvFile1.csv";
+    string csvFilePath2 = "./files/csvFile2.csv";
+    string[][] csvContent = [["1", "James", "10000"], ["2", "Nathan", "150000"], ["3", "Ronald", "120000"], ["4", "Roy",
+    "6000"], ["5", "Oliver", "1100000"]];
+
+    // Writes the given content string[][] to a CSV file.
+    check io:fileWriteCsv(csvFilePath1, csvContent);
+    // If the write operation was successful, then perform a read operation to read the CSV content as a string array of arrays.
```
Comment (praneesha) [inline examples/file-io/file_io.bal:85]:
```suggestion
    // If the write operation was successful, then, perform a read operation to read the CSV content as a string array of arrays.
```

File: examples/file-io/file_io.bal [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+    json readJson = check io:fileReadJson(jsonFilePath);
+    io:println(readJson);
+
+    // Initialize XML file path and content.
+    string xmlFilePath = "./files/xmlFile.xml";
+    xml xmlContent = xml `<book>The Lost World</book>`;
+
+    // Writes the given XML to a file.
+    check io:fileWriteXml(xmlFilePath, xmlContent);
+    // If the write operation was successful, then perform a read operation to read the XML content.
```
Comment (praneesha) [inline examples/file-io/file_io.bal:73]:
```suggestion
    // If the write operation was successful, then, perform a read operation to read the XML content.
```

File: examples/file-io/file_io.bal [hunk 18]
```diff
@@ -0,0 +1,97 @@
...
+            "address": {
+                "street": "Main",
+                "city": "94"
+            },
+            "codes": ["4", "8"]
+        }};
+
+    // Writes the given JSON to a file.
+    check io:fileWriteJson(jsonFilePath, jsonContent);
+    // If the write operation was successful, then perform a read operation to read the JSON content.
```
Comment (praneesha) [inline examples/file-io/file_io.bal:63]:
```suggestion
    // If the write operation was successful, then, perform a read operation to read the JSON content.
```

File: examples/file-io/file_io.bal [hunk 19]
```diff
@@ -0,0 +1,97 @@
...
+
+    // Writes the given array of lines to a file.
+    check io:fileWriteLines(textFilePath2, lines);
+    // If the write operation was successful, then perform a read operation to read the lines as an array.
+    string[] readLines = check io:fileReadLines(textFilePath2);
+    io:println(readLines);
+
+    // Writes the given stream of lines to a file.
+    check io:fileWriteLinesFromStream(textFilePath3, lines.toStream());
+    // If the write operation was successful, then perform a read operation to read the lines as a stream.
```
Comment (praneesha) [inline examples/file-io/file_io.bal:42]:
```suggestion
    // If the write operation was successful, then, perform a read operation to read the lines as a stream.
```

File: examples/file-io/file_io.bal [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+
+    // Writes the given string to a file.
+    check io:fileWriteString(textFilePath1, textContent);
+    // If the write operation was successful, then read the content as a string.
+    string readContent = check io:fileReadString(textFilePath1);
+    io:println(readContent);
+
+    // Writes the given array of lines to a file.
+    check io:fileWriteLines(textFilePath2, lines);
+    // If the write operation was successful, then perform a read operation to read the lines as an array.
```
Comment (praneesha) [inline examples/file-io/file_io.bal:36]:
```suggestion
    // If the write operation was successful, then, perform a read operation to read the lines as an array.
```

File: examples/file-io/file_io.bal [hunk 21]
```diff
@@ -0,0 +1,97 @@
...
+    // Initialize the text path and the content.
+    string textFilePath1 = "./files/textfile1.txt";
+    string textFilePath2 = "./files/textfile2.txt";
+    string textFilePath3 = "./files/textfile3.txt";
+    string textContent = "Ballerina is an open source programming language.";
+    string[] lines = ["The Big Bang Theory", "F.R.I.E.N.D.S", "Game of Thrones", "LOST"];
+
+    // Writes the given string to a file.
+    check io:fileWriteString(textFilePath1, textContent);
+    // If the write operation was successful, then read the content as a string.
```
Comment (praneesha) [inline examples/file-io/file_io.bal:30]:
```suggestion
    // If the write operation was successful, then, read the content as a string.
```

File: examples/file-io/file_io.description [hunk 22]
```diff
@@ -0,0 +1,3 @@
+// This example demonstrates how bytes can be read and written through the I/O APIs.<br/><br/>
```
Comment (praneesha) [inline examples/file-io/file_io.description:1]:
```suggestion
// This example demonstrates how bytes can be read and written using the I/O APIs.<br/><br/>
```

File: examples/file-io/file_io.out [hunk 23]
```diff
@@ -0,0 +1,25 @@
+# To run this sample, navigate to the directory that contains the
+# `.bal` file, and execute the `ballerina run` command below.
+# Note that, to run this examples successfully you should place an image in the file directory as follow.
```
Comment (praneesha) [inline examples/file-io/file_io.out:3]:
```suggestion
# Note that to run this examples successfully, you should place an image in the `files` directory as follows.
```

**Pull request merged**